File: update_environment_parameters.m

if (Gazebo_Sim == 1)
    Points360 = LaserSub.LatestMessage.ranges;
    Points360(Points360==Inf) = Lidar_Range;
else
    Points360 = Read_Lidar(X(:,Step_Counter), m2p, Lidar_Range, map_local);
    Points360_near = Read_Lidar(X(:,Step_Counter), m2p, Lidar_Range_near, map_local);
end
Points360 = Lidar_Range - Points360;
Points360_near = Lidar_Range_near - Points360_near;

Goal_Vector(:,Step_Counter) = [X_g(1,Step_Counter) - X(1,Step_Counter);X_g(2,Step_Counter) - X(3,Step_Counter)];
Goal_angle = atan2d(Goal_Vector(2,Step_Counter),Goal_Vector(1,Step_Counter));
Goal_Direction = Goal_angle - Robot.Heading;
Goal_Direction = (Goal_Direction + (360*(Goal_Direction<-180)) + (-360*(Goal_Direction>180)));
MF_Lid = MF_Lidar(Points360);
MF_Lid_near = MF_Lidar(Points360_near);

Phi2Goal = MF_Lidar_Angle - Goal_Direction;

SaturatedPHI2Goal = (Phi2Goal + (360*(Phi2Goal<-180)) + (-360*(Phi2Goal>180)));

weight_MF = gbellmf(SaturatedPHI2Goal,[bell_size, bell_coff, 0]);

Preference_MF = (weight_MF*0.7+0.3).*(1-MF_Lid);
Preference_MF_near = (weight_MF*0.7+0.3).*(1-MF_Lid_near);
% N_MF_Lid = Lidar_Range./min(Lidar_Range,max(MF_Lid,0.001));
% Preference_MF = (weight_MF*0.7+0.3).* N_MF_Lid;
Preference_MF = Preference_MF / max(Preference_MF);
Preference_MF_near = Preference_MF_near / max(Preference_MF_near);

File: update_environment_parameters.py
import numpy as np
import Read_Lidar
from init_params import *

# Read lidar data
Points360 = Read_Lidar(X[:, Step_Counter], X_g[:, Step_Counter], Lidar_Range, map)

# Compute goal vector
Goal_Vector[:, Step_Counter] = np.array([
    X_g[0, Step_Counter] - X[0, Step_Counter],
    X_g[1, Step_Counter] - X[2, Step_Counter]
])

# Calculate goal angle and direction
Goal_angle = np.degrees(np.arctan2(Goal_Vector[0, Step_Counter], Goal_Vector[0, Step_Counter]))
Goal_Direction = Goal_angle - Robot.Heading

# Compute membership function values for Lidar data
MF_Lid = MF_Lidar(Points360)

# Calculate angle differences
Phi2Goal = MF_Lidar_Angle - Goal_Direction

# Saturate angles to the range [-180, 180]
SaturatedPHI2Goal = Phi2Goal + (360 * (Phi2Goal < -180)) - (360 * (Phi2Goal > 180))

# Generate weight membership functions using generalized bell function
weight_MF = gbellmf(SaturatedPHI2Goal, [bell_size, bell_coff, 0])

# Calculate preference membership function
Preference_MF = (weight_MF * 0.7 + 0.3) * (1 - MF_Lid)
